gtsfm/loader/colmap_loader.py

```diff
@@ -0,0 +1,179 @@
+        print(f"Has {self._num_imgs} images")
+        # self.get_camera_intrinsics(0)
+
+        
+        # # read one image, to check if we need to downsample the images
+        # img = io_utils.load_image(self._image_paths[0])
+
+        # sample_h, sample_w = img.height, img.width
+        # if min(sample_h, sample_w) > self._max_resolution:
+        #     self._downsample_factor, self.target_h, self.target_w = img_utils.get_exact_downsample_factor(img, self._max_resolution)
+        # else:
+        #     self._downsample_factor = 1
+        self._downsample_factor = 4
+        
+
+    def __len__(self) -> int:
+        """The number of images in the dataset.
+
+        Returns:
+            the number of images.
+        """
+        return self._num_imgs
+
+    def get_image(self, index: int) -> Image:
+        """Get the image at the given index.
+
+        Args:
+            index: the index to fetch.
+
+        Raises:
+            IndexError: if an out-of-bounds image index is requested.
+
+        Returns:
+            Image: the image at the query index.
+        """
+
+        if index < 0 or index > self.__len__():
+            raise IndexError("Image index is invalid")
+
+        img = io_utils.load_image(self._image_paths[index])
+        img = img_utils.resize_image(img, new_height=img.height//self._downsample_factor, new_width=img.width//self._downsample_factor)
+        return img
+
+
+    def get_camera_intrinsics(self, index: int) -> Cal3Bundler:
+        """Get the camera intrinsics at the given index.
+
+        Args:
+            the index to fetch.
+
+        Returns:
+            intrinsics for the given camera.
+        """
+        if not self._use_gt_intrinsics:
+            # get intrinsics from exif
+            intrinsics = io_utils.load_image(self._image_paths[index]).get_intrinsics_from_exif()
+
+        else:
+            intrinsics = self._calibrations[index]
+
+        intrinsics = Cal3Bundler(
+            fx=intrinsics.fx() / self._downsample_factor,
+            k1=0.0,
+            k2=0.0,
+            u0=intrinsics.px() / self._downsample_factor,
+            v0=intrinsics.py() / self._downsample_factor,
+        )
+        return intrinsics
+
+
+    def get_camera_pose(self, index: int) -> Optional[Pose3]:
+        """Get the camera pose (in world coordinates) at the given index.
+
+        Args:
+            index: the index to fetch.
+
+        Returns:
+            the camera pose w_T_index.
+        """
+        if not self._use_gt_extrinsics:
+            return None
+
+        wTi = self._wTi_list[index]
+        return wTi
+
+
+    def validate_pair(self, idx1: int, idx2: int) -> bool:
+        """Checks if (idx1, idx2) is a valid pair.
+
+        Args:
+            idx1: first index of the pair.
+            idx2: second index of the pair.
+
+        Returns:
+            validation result.
+        """
+        return idx1 < idx2 and abs(idx1 - idx2) <= self._max_frame_lookahead
```

Comment:
no need to have `abs` because `idx1` is less than `idx2`.

Reply:
technically we would need to do `idx2 - idx1` instead, then. I think `abs()` makes it easier for the reader so they dont have to think about negative number edge cases at all

Reply:
True.